wallet/types/src/account_info.rs

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use common::address::pubkeyhash::PublicKeyHash;
+use common::chain::{OutPoint, Transaction};
+use common::primitives::Id;
+use crypto::key::extended::ExtendedPublicKey;
+use crypto::key::hdkd::child_number::ChildNumber;
+use crypto::key::hdkd::derivation_path::DerivationPath;
+use serialization::{Decode, Encode};
+use storage::HasPrefix;
+
+/// The account id is described by it's public key hash or a random value for
+/// non deterministic accounts
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Encode, Decode)]
+pub struct AccountId(PublicKeyHash);
+
+impl AccountId {
+    pub fn new_from_xpub(pub_key: &ExtendedPublicKey) -> Self {
+        let pk = pub_key.clone().into_public_key();
+        let pkh = PublicKeyHash::from(&pk);
+        AccountId(pkh)
+    }
+}
+
+/// Account metadata that contains information like from which master key it was derived from
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Encode, Decode)]
+pub enum AccountInfo {
+    #[codec(index = 0)]
+    Deterministic(DeterministicAccountInfo),
+}
+
+/// Struct that holds information for account addresses
+#[derive(Debug, Default, Clone, PartialEq, Eq, PartialOrd, Ord, Encode, Decode)]
+pub struct KeychainUsageState {
+    /// Last used address index. An address might not be used so the corresponding entry would be
+    /// None, otherwise it would be that last used ChildNumber
+    last_used: Option<ChildNumber>,
```

Comment:
How are gaps handled here? Is this just to present the user with a fresh receiving address and leaving the unused child numbers in the `[0, last_used)` interval alone?

Tip: not sure it's an improvement but a thing to consider. Instead of `last_used` and `last_issued`, you can have `first_unused`, `next_to_issue` (or something like that) and avoid the `Option`. This is conceptually similar how ranges work, where the upper bound is not part of the range. E.g. the range of used addresses is `0..first_unused`.

Reply:
I will tackle this in an upcoming PR.